### DISCUSSION TITLE
GitHub Models & OPs Report Update

### DISCUSSION THREAD

**Discussion by ayerofieiev-tt** [tenstorrent/pytorch2.0_ttnn#570] [General]
# GitHub Models & OPs Report Update
1. Modes on the main page are now sorted by status
2. Instead of PCC true/false we now output specific value (might need to truncate)
3. New column added with info on how many times data was transferred to host during an operation run

I am happy about all 3 changes, but the last one is a jewel. Here is see why.

## First, why does it matter?
Now we can actually tell if a model runs e2e on device.
Now as this data is on the surface, we can dig dipper and improve.

## aten.embedding 
I looked into BERT and saw that `aten.embedding` calls generates 2 unexpected fallbacks to host after conversion to TT-NN. 
<img width="793" alt="Screenshot 2024-12-08 at 8 56 07 PM" src="https://github.com/user-attachments/assets/aa9fc8e4-f052-4fcc-9dfc-1d4498ac2cc6">
Wow! Thats not good. 

Here is how conversion for this OP looks like: an embedding call and a conditional to_layout call at the end.
```py
if node.target == torch.ops.aten.embedding.default:
    tiled = args[1].meta.get("val")
    tiled = False if tiled is None else tiled.size()[-1] % ttnn.TILE_SIZE == 0
    layout = TtnnTileLayout() if tiled else TtnnRowMajorLayout()
    tensor = g.call_function(ttnn.embedding, (args[1], args[0]), {"layout": layout})
    return tensor if tiled else g.call_function(ttnn.to_layout, (tensor, TtnnTileLayout()))
```
We expect that in an isolated test a compiled module:
1. Converts 2 inputs from Torch to TT-NN, sending them to device in Tile layout
2. Calls embedding
3. Brings result to host and converts it to torch
So only a single device --> host transition is expected, while report tells us that we got 3 calls!

Lets run the isolated test and see whats actually going on:
```
pytest tests/autogen_op/BERT/test_BERT_aten_embedding_default.py -s
```
Compiled module looks like this (find this in run logs):
```
4: def forward(self, arg0_1, arg1_1):
5:     ttnn_from_torch = ttnn_decorators_ttnn_from_torch(arg1_1, layout = ttnn_TILE_LAYOUT, device = ttnn_Specified_Device, dtype = ttnn_uint32);  arg1_1 = None
6:     ttnn_from_device = ttnn_decorators_ttnn_from_device(ttnn_from_torch);  ttnn_from_torch = None
7:     ttnn_to_layout = ttnn_decorators_ttnn_to_layout(ttnn_from_device, ttnn_ROW_MAJOR_LAYOUT);  ttnn_from_device = None
8:     ttnn_to_device = ttnn_decorators_ttnn_to_device(ttnn_to_layout, device = ttnn_Specified_Device);  ttnn_to_layout = None
9:     ttnn_from_torch_1 = ttnn_decorators_ttnn_from_torch(arg0_1, layout = ttnn_TILE_LAYOUT, device = ttnn_Specified_Device, dtype = ttnn_bfloat16);  arg0_1 = None
10:     ttnn_from_device_1 = ttnn_decorators_ttnn_from_device(ttnn_from_torch_1);  ttnn_from_torch_1 = None
11:     ttnn_to_layout_1 = ttnn_decorators_ttnn_to_layout(ttnn_from_device_1, ttnn_ROW_MAJOR_LAYOUT);  ttnn_from_device_1 = None
12:     ttnn_to_device_1 = ttnn_decorators_ttnn_to_device(ttnn_to_layout_1, device = ttnn_Specified_Device);  ttnn_to_layout_1 = None
13:     ttnn_embedding = ttnn_decorators_ttnn_embedding(ttnn_to_device, ttnn_to_device_1, layout = ttnn_TILE_LAYOUT);  ttnn_to_device = ttnn_to_device_1 = None
14:     ttnn_to_torch = ttnn_decorators_ttnn_to_torch(ttnn_embedding);  ttnn_embedding = None
15:     return (ttnn_to_torch,)
```
In other words, this is what I see:

> :white_check_mark: line 5: take one input from torch, onto device, in tile layout
> :scream: line 6: move that tensor from device. I am confused why this happens
> :scream: line 7: change layout to row major
> :scream: line 8: send back to device
> :white_check_mark: line 9: take another input from torch into device in tile layout
> :scream: line 10: take it from device to host
> :scream: line 11: change layout to row major
> :scream: line 12: send back to device
> :white_check_mark: line 13: call the actual op
> :white_check_mark: line 14: convert result to torch

We clearly see those 2 unexpected device-->host transitions here.
They are not happening inside the embedding call, but caused by a compiler generated code.
Even if embedding can't intake tiled inputs, we still can choose the right layout and eliminate those transitions in both cases.
Yay!

## aten.split
aten.split is another operation which shows redundant transfers to host.
<img width="652" alt="Screenshot 2024-12-08 at 8 55 28 PM" src="https://github.com/user-attachments/assets/c97f3554-8e17-4141-bd7d-36b8a99e43c7">

This is the code compiler produces for an isolated module:
```
4: def forward(self, arg0_1):
5:     ttnn_from_torch = ttnn_decorators_ttnn_from_torch(arg0_1, layout = ttnn_TILE_LAYOUT, device = ttnn_Specified_Device, dtype = ttnn_bfloat16);  arg0_1 = None
6:     ttnn_to_layout = ttnn_decorators_ttnn_to_layout(ttnn_from_torch, layout = ttnn_ROW_MAJOR_LAYOUT);  ttnn_from_torch = None
7:     ttnn_split = ttnn_decorators_ttnn_split(ttnn_to_layout, 2, 2);  ttnn_to_layout = None
8:     getitem = ttnn_split[0]
9:     getitem_1 = ttnn_split[1];  ttnn_split = None
10:     ttnn_to_torch = ttnn_decorators_ttnn_to_torch(getitem);  getitem = None
11:     ttnn_to_torch_1 = ttnn_decorators_ttnn_to_torch(getitem_1);  getitem_1 = None
12:     return (ttnn_to_torch, ttnn_to_torch_1)
```
Compiler adds conversion from Tile to Row Major layout, which might not be necessary. But otherwise there are no unexpected data transfers to host generated by compiler. Now graph tracing comes into light, kudos to @dmakoviichuk-tt 

Simplified, [this is how we trace](https://github.com/tenstorrent/pytorch2.0_ttnn/blob/e7b9b472e76ecb33d6db98cb112fe0693e7fee25/tools/templates/aten_test.tmpl#L58) a module:
```py
ttnn.graph.begin_graph_capture()
result_after = m.forward(*input_args, **input_kwargs)
trace = ttnn.graph.end_graph_capture()
call_stack = ttnn.graph.extract_calltrace(trace)
to_host_count = sum(["Tensor::cpu" in str(node) for node in call_stack])
```
The method is not perfect, but it gives us the insight. Here is how the call trace looks like (with some handcrafted additionss:
```
👉 Transfer input to host 
tt::tt_metal::detail::convert_python_tensor_to_tt_tensor
ttnn::to_layout
    Tensor::pad
    Tensor::to
    ttnn::reshape
Tensor::to

👉 Switch to Row Major layout
ttnn::to_layout
    ttnn::untilize_with_unpadding
        ttnn::prim::old_infra_device_operation
            UntilizeWithUnpadding
                tt::tt_metal::create_device_tensor
    ttnn::reshape

👉 Actual call of split
ttnn::split
    ttnn::reshape
        Tensor::reshape
    Tensor::cpu <<<<------ oops, here you are ❗️
    👉 Slice chunk 0
    ttnn::slice
        ttnn::to_layout
        Tensor::unpad
        ttnn::to_layout
        ttnn::reshape
    ttnn::reshape
        Tensor::reshape
    Tensor::pad_to_tile
        Tensor::pad
    Tensor::to
    Tensor::to

    👉 Slice chunk 1
    ttnn::slice
        ttnn::to_layout
        Tensor::unpad
        ttnn::to_layout
        ttnn::reshape
    ttnn::reshape
        Tensor::reshape
    Tensor::pad_to_tile
        Tensor::pad
    Tensor::to
    Tensor::to

👉 Transfers chunk 0 to host
Tensor::cpu
Tensor::to
tt::tt_metal::detail::convert_tt_tensor_to_torch_tensor

👉 Transfers chunk 1 to host
Tensor::cpu
Tensor::to
tt::tt_metal::detail::convert_tt_tensor_to_torch_tensor
```

So it is `ttnn::split` that adds a transfer to host. It needs some work.

[Shape from bert](https://github.com/tenstorrent/pytorch2.0_ttnn/blob/main/docs/models/BERT/input_variations.md#atensplittensor)
[All shapes](https://github.com/tenstorrent/pytorch2.0_ttnn/blob/main/docs/operations/aten.split.Tensor.md)
Don't get surprised to see a negative host transfers count - those are for fallbacks - we expected to see N transfers and got 0, so host stat is `-N`.

**Comment by swimdi** [comment]
I think the embedding issue is related with this

https://github.com/tenstorrent/pytorch2.0_ttnn/blob/e7b9b472e76ecb33d6db98cb112fe0693e7fee25/torch_ttnn/passes/lowering/add_data_move_pass.py#L405-L422

I originally just want `ttnn.from_torch(device, ROW_MAJOR, )`, but the CI of `perceiver_io` will failed and can't merged, so I kept it `from_torch(device, TILE) => from_device => to_layout(ROW_MAJOR) => to_device`
